Some more screen changes, add tests

psychon opened the pull request:
Hi everyone,

this adds a new API to fake screen additions and removals. Based on this, a new functional test is added that does actually test this. The rest are various fixes.

However, since tags are leaked (hrm, `tag:delete()` should clean up better), 17dab5446ef59933601dc4bf9bd706cc793a6f0d is necessary to make the test pass. Also, https://github.com/awesomeWM/awesome/commit/17dab5446ef59933601dc4bf9bd706cc793a6f0d#diff-f9bfa3f1498dea81c7ec01763ba182a8R30 is really ugly.

This last commit is the WIP part of this PR and I'd be nice to hear some ideas from you what to do about this. If you want, I can remove the "add a new test" and the WIP commits and the rest can be merged.

psychon commented:
Oh and cb6ce838a3fe6f8b53ecec22760ae4ed7ca9a504 might fix #861!
@Elv13 Any particular reason why you use `GtkApplication` besides "all the lgi examples do so as well"?

Elv13 commented:
> Any particular reason why you use GtkApplication besides "all the lgi examples do so as well"?

It is a copy pasted (and trimmed) LGI example

Elv13 commented on lib/awful/tag.lua:
```diff
@@ -1355,6 +1355,32 @@ capi.screen.add_signal("tag::history::update")
+    local fallback = nil
+    for other_screen in capi.screen do
+        if #other_screen.tags > 0 then
+            fallback = other_screen.tags[1]
```
This should be done in a request::screen handler. Tyrannical may want to move "volatile" tags to existing screens instead of just deleting them. This is also true for the `:delete()` above. The default handler should delete the tags. This make total sense for the default config where tags are less or more used as virtual workspaces.

psychon replied:
Oh, I totally forgot to say "I don't like this part at all and I'm open for suggestions".
How about some "special" signal just for "this tag must go away, do something suitably nice" that is emitted at the beginning of this function? That way tyrannical can hook into that signal and do everything it needs to do and the rest of this function won't be necessary.

Thanks for this idea!

So, what do you think about:
```
screen.connect_signal("removed", function(s)
    for _, t in pairs(s:tags) do
        t:emit_signal("delete-for-screen-removal")
    end
    [Now here the rest of the code]
end)
```
(Also: Ideas for a good name for the signal?)

Elv13 commented on objects/screen.c:
```diff
@@ -1016,6 +1016,57 @@ luaA_screen_count(lua_State *L)
+ * @function fake_add
+ */
+static int
+luaA_screen_fake_add(lua_State *L)
```
Could there be a fake_resize too? It would be useful to be able to split Xephyr screen into "virtual sub screens"[1] for the tests (or use the Xinerama support and just virtually enable and disable the other screens. Also, this is actually useful for people using buggy hardware (like the Radeon HD5450[2]), MOSAIC or other unsupported ways to define screens.

[1] Virtually resize the real screen in 2, then fake_add the second part of the real screen as an imaginary second screen.
[2] That one support 3 screens per card, but expose them as 2 screens because of an (hardware?) bug.

psychon replied:
Yup, that's planned, but only after "just disabling" screens works. One problem at a time. :-)

psychon commented:
No longer a WIP, I hope. The diff to the previous version of this PR is https://gist.github.com/psychon/676ec08155cee1dd44aaedcd56342471 (but hand-edited, so the patch might be invalid).

Does this look mergable?

Elv13 commented on lib/awful/tag.lua:
```diff
@@ -1350,11 +1352,35 @@ capi.tag.add_signal("property::urgent")
+capi.screen.connect_signal("removed", function(s)
+    -- First give other code a chance to clean up
+    for _, t in pairs(s.tags) do
+        t:emit_signal("removal-pending")
```
I know I could implement `request::tag` myself using this signal, but I would prefer it to be integrated into the "default" sequence.

First, tags should emit a `request::screen` without any handler by default. If the screen changed, then return. After that, the client should each emit `request::tag`. The default handler will already do less or more what's your doing here. This way, the tag can be salvaged with its layout and all and just moved to another screen.

psychon replied:
So you are saying:

1. I should rename `removal-pending` to `request::screen` (well, how about `request::new_screen`?)
2. I should do another loop afterwards with `for _, c in ipairs(s.clients) do c:emit_signal("request::tag") end` (how about `request::new_tag`?)

(And after the above, I keep the code that I already have)

Right?

Elv13 commented:
`removal-pending` is still required if `request::screen` end up with no screen. If no handler decided to move the tag, then the code should go ahead and start to delete it.

> c:emit_signal("request::tag") end (how about request::new_tag?)

`request::tag` already exist and does that. As implemented, there is already an hint. So "reason" should be "screen_removed" or somehting like that. Here is how it is currently used:

```lua
 c:emit_signal("request::tag", nil, {reason="rules"})
```

psychon commented:
Rebased on master for resolving a conflict and added https://github.com/awesomeWM/awesome/pull/868/commits/5b1f8cd35c2b6ea88778238ef13b1e96ef5f32c1 to hopefully make Elv13 happy

Elv13 commented:
:+1:
